### PR TITLE
Restructure Documents to support bulk embedding

### PR DESCRIPTION
**Pull request by tomusher**
Forgive the large PR here - this started out as me implementing bulk embedding and then going down a spiral of refactoring things because the existing structure made bulk operations quite difficult.

This change refactors how Documents are generated and used in order to:

- Decouple them from content types (resolves https://github.com/wagtail/wagtail-vector-index/issues/73)
- Support bulk embeddings (resolving https://github.com/wagtail/wagtail-vector-index/issues/84)
- Simplify the code base and make individual components more testable.
- Hopefully make it easier to filter searches by querysets.

In more detail:

- Where we used to have a `Document` dataclass that was converted back and forth to `Embedding` model instances, there is now only a `Document` Django model. `Documents` represent a chunk of something to be stored/queried on by a storage provider.
- `DocumentConverters` are responsible for converting *something* to Documents, and Documents back to their relevant *something*.
- A `Document` no longer has references to objects or content types - each document can have list of `object_keys` which are an arbitrary representation of the thing it relates to. A `DocumentConverter` knows how to both generate keys for the object it's working with, and retrieve a relevant object based on a key.
- In the case of the `ModelDocumentConverter`, this can create `Documents` based on a `ModelKey`, and (relatively) efficiently get all Django models that correspond to a list of `Documents`.
- `DocumentConverters` are now composed of separate operators that can be individually specified/overridden to make it easier to test and reason about each operation a converter is involved with.

### CODE REVIEW

File: src/wagtail_vector_index/storage/django.py [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+- The EmbeddableFieldsMixin, which is a mixin for Django models that lets user define which fields should be used to generate embeddings
+- The EmbeddableFieldsVectorIndexMixin, which is a VectorIndex mixin that expects EmbeddableFieldsMixin models
+- The EmbeddableFieldsDocumentConverter, which is a DocumentConverter that knows how to convert a model instance using the EmbeddableFieldsMixin protocol to and from a Document
+"""
+
+ModelLabel: TypeAlias = str
+ObjectId: TypeAlias = str
+
+
+def batched(iterable, n):
```
Comment (zerolab) [inline src/wagtail_vector_index/storage/django.py:65]: drive-by suggestion: test if `itertools` has the `batched` attribute (Python 3.12+) and use that before using our implementation. https://docs.python.org/3.12/library/itertools.html#itertools.batched

Comment (tomusher) [reply src/wagtail_vector_index/storage/django.py]: Thanks @zerolab - good call!

File: src/wagtail_vector_index/storage/django.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+        Same as `_get_models_by_key`, but async.
+        """
+        objects_by_key: dict[ModelKey, models.Model] = {}
+        for model_label, keys in keys_by_model_label.items():
+            model_class = ModelFromDocumentOperator._model_class_from_label(model_label)
+            model_objects = model_class.objects.filter(
+                pk__in=[key.object_id for key in keys]
+            )
+            objects_by_key.update(
+                {ModelKey.from_instance(obj): obj for obj in model_objects}
```
Comment (jacobtoppm) [inline src/wagtail_vector_index/storage/django.py]: Is this meant to be evaluated in an async-friendly way?

File: src/wagtail_vector_index/storage/models.py [hunk 3]
```diff
@@ -5,7 +5,7 @@
 from django.db.models import Q
 
 
-class DocumentQuerySet(models.QuerySet):
+class DocumentManager(models.Manager):
```
Comment (jacobtoppm) [inline src/wagtail_vector_index/storage/models.py:8]: Wouldn't the QuerySet derived manager be preferable here for allowing chaining custom operations + default queryset operations regardless of order? Not a big deal, but I'm not sure I get this change

**Review by jacobtoppm — COMMENTED** [review]
Generally looks great! Added a couple questions/nitpicks

File: src/wagtail_vector_index/storage/django.py (hunk 2 above)
Comment (jacobtoppm) [reply src/wagtail_vector_index/storage/django.py]: Ah wait, you did this in a later commit.

File: src/wagtail_vector_index/storage/models.py (hunk 3 above)
Comment (tomusher) [reply src/wagtail_vector_index/storage/models.py:8]: Yes definitely preferable but I ran in to typing issues using the `QuerySet` derived manager.

I think this is resolved in django-stubs by https://github.com/typeddjango/django-stubs/issues/738, but not in django-types which we're using in this project.

File: src/wagtail_vector_index/storage/django.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+        objects_by_key = self._get_models_by_key(keys_by_model_label)
+
+        yield from self._get_deduplicated_objects_generator(documents, objects_by_key)
+
+    async def abulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> AsyncGenerator[models.Model, None]:
+        """A copy of `bulk_from_documents`, but async"""
+        # Force evaluate generators to allow value to be reused
+        documents = tuple(documents)
```
[the remark this answers is not in the document]
Comment (tomusher) [reply src/wagtail_vector_index/storage/django.py]: Yeah you're right - I've changed these methods to accept Sequences to prevent the need to force evaluate. You could still conceivably pass a Sequence that can't be evaluated in an async context but I'm not sure if there's a way to guard against that - any ideas welcome!

[the remark this answers is not in the document]
Comment (jacobtoppm) [reply src/wagtail_vector_index/storage/django.py]: Looks good! Honestly I can't think of a great way to impose that restriction either - ie get rid of the chance of `.only` etc deferring fields we want - without forcing it to no longer be a sequence of documents, and instead plain dataclasses... which seems like overkill.

**Review by jacobtoppm — APPROVED** [review]
